### ISSUE TITLE
Lassoing links no longer working properly

### ISSUE THREAD

**Issue by jeff-harp** [cpriest/SnapLinksPlus#308]
This may be a duplicate of #302.

Ubuntu 18.04LTS
Firefox 74.0 (just upgraded from 73.0.1)
Snap Links 3.1.7
Select set to Ctrl & Shift + Left Click

I am configured for manual upgrade of Firefox and add-ons, so I know nothing changed other than the Firefox upgrade.

My normal usage of Snap Links is to copy multiple links for pasting into other applications.

Here's a screenshot showing what happens when I try to lasso the 3 source package links on right side of page at https://packages.ubuntu.com/bionic/firefox

![lasso](https://user-images.githubusercontent.com/47549560/76625971-adea9180-650e-11ea-9b0a-7fc5f4b4822a.png)

As you can see, a lot of text is highlighted. What gets highlighted varies greatly depending on where pointer is when lassoing starts. Not consistent at all as far as I can tell. Once in a while I can lasso without any unwanted highlighting.

Although annoying, I wouldn't mind the highlighting if the links got copied properly, but they are not being copied sometimes (maybe 10% of the time), even though it shows while lassoing that 3 links are selected. Not consistent at all, and not easy to deliberately not copy the links when trying to reproduce. When not copied, buffer contains whatever was copied last.

**Comment by cpriest** [comment]
I would bet this has to do with some new permission stuff that's gone on regarding the clipboard.

Is the text highlighting something that's new or has that always been present?  That might go away but using something other than Ctrl & Shift, one of those keys may be "unpreventable" in Firefox.

I'll look into this as soon as I am able.

**Comment by jeff-harp** [comment]
Thanks for the quick response.

The text selection/highlighting is new. In the past, there would be some reasonable highlighting taking place in close proximity to the selection box. But now it's just selecting things all over the place, with no obvious relationship to where the selection box is. The example page I mentioned above is the most extreme case I've seen. Other pages can be more or less "normal".

I switched from "Ctrl & Shift + Left Click" to "Ctrl & Alt + Left Click" to see what would happen. The unwanted selection behaviour goes away. So I'll just use this new key combo from now on.

It's interesting to note that the strange selection behaviour in Firefox with "Ctrl & Shift" continues, which suggests that it has nothing to do with Snap Links. Perhaps it would be wise to make that key combo unavailable in Firefox for Snap Links, or at least warn users (if you don't already) that they may see some strange selection behaviour when using it.

The "other" problem I mentioned where nothing is being copied is now easily reproducible for me, and occurs when you collapse a selection box horizontally to the point where it "crosses over" and expands on the opposite side. If I do that twice, even though it says it has "N links", it does not, and if I paste I get whatever was in the buffer (clipboard) before. If I cross over only once, copying works properly. I did not test this in the vertical direction, nor did I test it with more than two cross overs occurring.

So there are really two issues here. Should I close this issue and open two new ones in its place? I apologize for the confusion. I should have done more testing before my initial post.

**Comment by cpriest** [comment]
Hmm, this ticket should be fine.  Did you create a test case HTML file showing the issue by chance?  If so, please upload it to this ticket, that will help a lot.

**Comment by jeff-harp** [comment]
I revised my last post to correct a mistake regarding when the failure to copy links occurs for me. It happens on the first crossover of the selection box.

Here's some HTML that shows step-by-step how to reproduce the issue:

```html
<!DOCTYPE html>
<html>
   <head>
      <title>Snap Links Plus selection issue</title>
      <meta charset="UTF-8">
      <style>
         body {
            margin: 50px;
         }
         body p {
            white-space: pre-line;
            line-height: 1.5;
         }
         .link_box {
            display: inline-block;
            position: relative;
            /* border: 1px solid black; */
         }
         .link_box a {
            display: block;
            line-height: 1.5;
         }
         .target {
            width: 25px;
            height: 25px;
            border-radius: 50%;
            display: flex;
            position: absolute;
            justify-content: center;
            align-items: center;
            background-color: red;
            color: white;
            font-family: sans-serif;
            font-weight: normal;
            text-align: center;
            cursor: default;
         }
         .targetA {
            left: 40px;
            top: -40px;
         }
         .targetB {
            left: 40px;
            top: 110px;
         }
         .targetC {
            left: 80px;
            top: 110px;
         }
         .targetD {
            left: 0px;
            top: 110px;
         }
      </style>
   </head>
   <body>
      <div class="link_box">
            <div class="target targetA">A</div>
            <div class="target targetB">B</div>
            <div class="target targetC">C</div>
            <div class="target targetD">D</div>
            <a href="https://github.com/cpriest/SnapLinksPlus/issues/308">Issue #308</a>
            <a href="https://cpriest.github.io/SnapLinksPlus/#/">github.io</a>
            <a href="https://github.com/cpriest/SnapLinksPlus">github.com</a>
            <a href="https://addons.mozilla.org/en-US/firefox/addon/snaplinksplus/">addons.mozilla.org</a>
      </div>
      <p>

         To reproduce this issue, follow these steps:

         1) Make sure Snap Links Plus (SLP) is configured to display the number of links selected.
         2) Double-click on word "issue" above and press Ctrl-C so clipboard content is known for subsequent comparison.
         3) Position pointer on "A".
         4) Initiate selection (e.g., Ctrl &amp; Alt + Left Click)
         5) Move pointer to "B".
         6) Move pointer to "C".
         7) Move pointer to "D".
         8) SLP should show 4 links selected.
         9) Complete selection - release mouse button and keys.
         10) Paste clipboard. Should be the word "issue", instead of the links that were selected.
      </p>
   </body>
</html>

**Comment by remi-garcia** [comment]
I was able to reproduce your last bug (your "HTML step-by-step how to reproduce" was really helpful!).

However I can't reproduce the "highlighted text" part.

I'll look into the copy part asap.

**Comment by cpriest** [comment]
I'm sorry I hadn't had a chance to test this out before.  This is expected behavior because holding alt + ctrl to start the drag is all that's necessary.  After the lasso has been started, alt + ctrl takes on a different meaning.

One of them specifies to copy the links to the clipboard.  This is only if it is being held down when the mouse button is released.

So the simple answer here is to use the ctrl + alt to start the lasso, then release them (but not the left click) and finish the lasso by releasing the left click (while not holding down any modifiers).

This functionality is probably going to be changing soon back to the way it was years ago before I took Snap Links over for the project.  It was much simpler then.

Let me know if this is still an issue beyond expected functionality, otherwise I'll close this ticket, thanks @jeff-harp @remi-garcia

**Comment by remi-garcia** [comment]
I can't reproduce any of the bugs.
